### PR TITLE
43 create simple ego pedestrian

### PR DESCRIPTION
**Pull request by JuliusMiller**
This PR introduces a new agent for the environment: The Ego Pedestrian.

The current goal of the pedestrian is to move to the robot's position and collide with it.

New:
PedestrianEnv, PedestrianState, UnicycleDrive, train & debug files

Adjusted:
- eval, tb_logging: for collection of training data
- env_config, env_util: methods for pedestrian creation
- reward: method for reward calculation
- robot_env: changes for new gym version
- occupancy: rename to agent, extended  occupancy for pedestrian
- svg_map_parser: fix spawn point problem
- sim_view: new key shortcuts, visualization of pedestrian
- robot_state: robot_robot collision is not implemented
- range_sensor: add method to detect the robot as ego_pedestrian
- simulator: extend simulator for pedestrian

New Tests:
- PedestrianEnv
- EnvMetrics
- lidar_sensor
- map_test
- sb3: adjustment for new gym version
- unicycle_drive
Fixes: #43 #52 

## Release Notes

- **New Features**
  - Introduced a comprehensive configuration file for automated code reviews and project management integrations.
  - Enabled a free tier for the code review system with automatic reviews and high-level summaries.
  - Added support for multiple code analysis tools, enhancing the review process.
  - Enhanced the simulator to support pedestrian dynamics alongside robots, improving interaction realism.
  - Implemented a training script for a reinforcement learning model focused on pedestrian dynamics.
  - Added a new class for pedestrian state management to facilitate collision detection and goal tracking.

- **Bug Fixes**
  - Improved readability in test code for better logical separation, enhancing maintainability.

- **Tests**
  - Enhanced test structure for improved clarity without altering functionality.

- **Documentation**
  - Updated documentation to reflect the new configuration settings and features in the code review system.

### CODE REVIEW

File: tests/env_test.py [hunk 1]
```diff
@@ -27,6 +28,19 @@ def test_can_simulate_with_pedestrians():
...
+def test_ego_ped_env():
+    env = PedestrianEnv()
+    assert env is not None
+
+    total_steps = 1000
+    env = RobotEnv()
+    env.reset()
+    for _ in range(total_steps):
+        rand_action = env.action_space.sample()
+        _, _, done, _, _ = env.step(rand_action)
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply tests/env_test.py:44]: Adjustment made to better test the pedestrian env

File: examples/ego_ped_example.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    obs = env.reset()
+
+    logger.info("Simulating the random policy.")
+    for _ in range(10000):
+        action_ped = env.action_space.sample()
+        obs, _, done, _ , _= env.step(action_ped)
+        env.render()
+
+    env.reset()
+    env.exit()
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply examples/ego_ped_example.py:43]: Included the done condition

File: examples/ego_ped_example.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    converter = SvgMapConverter(svg_file)
+    return converter.map_definition
+
+def get_file():
+    """Get the latest recorded file."""
+
+    filename = max(
+        os.listdir('recordings'), key=lambda x: os.path.getctime(os.path.join('recordings', x)))
+    return Path('recordings', filename)
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply examples/ego_ped_example.py:59]: Is not needed

File: tests/unicyle_drive_test.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    assert speed == 0 # assert stopped
+
+    motion.move(state, deceleration_action, 1.0)
+    speed = state.velocity
+    assert speed == -1 # assert moving backwards
+
+    motion = UnicycleMotion(UnicycleDriveSettings(allow_backwards=False))
+    motion.move(state, deceleration_action, 1.0)
+    speed = state.velocity
+    assert speed == 0 # assert not moving backwards
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply tests/unicyle_drive_test.py:76]: Done

File: robot_sf/nav/svg_map_parser.py [hunk 5]
```diff
@@ -279,6 +273,25 @@ def _info_to_mapdefintion(self) -> MapDefinition:
...
+        return spawn, goal
+
+def convert_map(svg_file: str):
+    """Create MapDefinition from svg file."""
+
+    logger.info("Converting SVG map to MapDefinition object.")
+    logger.info(f"SVG file: {svg_file}")
+
+    converter = SvgMapConverter(svg_file)
+    return converter.map_definition
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/nav/svg_map_parser.py:297]: Done

File: scripts/training_ped_ppo.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        robot_model = PPO.load("./model/run_043", env=None)
+
+        env_config = PedEnvSettings(
+            map_pool=MapDefinitionPool(map_defs={"my_map": map_definition}),
+            sim_config=SimulationSettings(difficulty=0, ped_density_by_difficulty=[0.02]),
+            robot_config=BicycleDriveSettings(radius=0.5, max_accel=3.0, allow_backwards=True)
+        )
+        env_config.sim_config.ped_density_by_difficulty = ped_densities
+        env_config.sim_config.difficulty = difficulty
+        return PedestrianEnv(env_config, robot_model=robot_model)
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply scripts/training_ped_ppo.py:35]: Done

File: scripts/debug_ped_policy.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    model = PPO.load(filename, env=env)
+
+    obs = env.reset()
+    ep_rewards = 0
+    for _ in range(10000):
+        if isinstance(obs, tuple):
+            action, _ = model.predict(obs[0], deterministic=True)
+        else:
+            action, _ = model.predict(obs, deterministic=True)
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply scripts/debug_ped_policy.py:52]: Done

File: scripts/debug_ped_policy.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    map_definition = convert_map("maps/svg_maps/debug_02.svg")
+    robot_model = PPO.load("./model/run_043", env=None)
+
+    env_config = PedEnvSettings(
+        map_pool=MapDefinitionPool(map_defs={"my_map": map_definition}),
+        sim_config=SimulationSettings(difficulty=0, ped_density_by_difficulty=[0.02]),
+        robot_config=BicycleDriveSettings(radius=0.5, max_accel=3.0, allow_backwards=True)
+    )
+    env_config.sim_config.ped_density_by_difficulty = ped_densities
+    env_config.sim_config.difficulty = difficulty
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply scripts/debug_ped_policy.py:28]: Done

File: robot_sf/gym_env/env_config.py [hunk 9]
```diff
@@ -60,3 +65,34 @@ def robot_factory(self) -> Union[DifferentialDriveRobot, BicycleDriveRobot]:
         else:
             raise NotImplementedError(
                 f"unsupported robot type {type(self.robot_config)}!")
+
+
+@dataclass
+class PedEnvSettings(EnvSettings):
+    """
+    Data class to hold environment settings for a simulation.
+    """
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/gym_env/env_config.py:74]: Done

File: tests/map_test.py [hunk 10]
```diff
@@ -48,5 +51,25 @@ def test_is_collision_with_obstacle_segment_outside_circle():
...
+def test_proximity_point():
+    fixed_point = (50, 50)
+    lower_bound = 15
+    upper_bound = 20
+    env_config = PedEnvSettings()
+    svg_file = "maps/svg_maps/debug_03.svg"
+    map_def = convert_map(svg_file)
+    _sim = init_ped_simulators(env_config, map_def)[0]
+    new_point = _sim.get_proximity_point(fixed_point, lower_bound=lower_bound, upper_bound=upper_bound)
+    assert lower_bound <= dist(fixed_point, new_point) <= upper_bound
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply tests/map_test.py:74]: Done

File: robot_sf/gym_env/env_config.py [hunk 11]
```diff
@@ -60,3 +65,34 @@ def robot_factory(self) -> Union[DifferentialDriveRobot, BicycleDriveRobot]:
...
+    def __post_init__(self):
+        """
+        Check if any of the properties are not initialized (None) and raise an
+        error if so.
+        """
+        if not self.sim_config or not self.lidar_config \
+                or not self.robot_config or not self.map_pool \
+                or not self.ego_ped_config:
+            raise ValueError('Please make sure all properties are initialized!')
+
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/gym_env/env_config.py:86]: Done

File: robot_sf/tb_logging.py [hunk 12]
```diff
@@ -49,3 +49,53 @@ def _on_step(self) -> bool:
...
+            self.writer.add_scalar("metrics/robot_at_goal_rate",
+                                   self.metrics.robot_at_goal_rate, self.num_timesteps)
+            self.writer.add_scalar("metrics/robot_obstacle_collision_rate",
+                                   self.metrics.robot_obstacle_collision_rate, self.num_timesteps)
+            self.writer.add_scalar("metrics/robot_pedestrian_collision_rate",
+                                   self.metrics.robot_pedestrian_collision_rate, self.num_timesteps)
+            self.writer.add_scalar("metrics/avg_distance_to_robot",
+                                   self.metrics.route_end_distance, self.num_timesteps)
+            self.writer.flush()
+        return True # info: don't request early abort
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/tb_logging.py:101]: Done

File: robot_sf/ped_ego/pedestrian_state.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+            for collision detection.
+        sensors (SensorFusion): Object for managing sensor data and fusion.
+        d_t (float): The simulation timestep duration.
+        sim_time_limit (float): The maximum allowed simulation time.
+
+    Additional attributes with default values initialized to False or 0 representing various
+    conditions such as collision states, timeout condition, simulation time elapsed,
+    and timestep count.
+    These are updated during the simulation.
+    """
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/ped_ego/pedestrian_state.py:33]: Done

File: robot_sf/ped_ego/pedestrian_state.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    robot_occupancy: ContinuousOccupancy
+    ego_ped_occupancy: EgoPedContinuousOccupancy
+    sensors: SensorFusion
+    d_t: float
+    sim_time_limit: float
+    episode: int = field(init=False, default=0)
+    is_at_goal: bool = field(init=False, default=False)
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/ped_ego/pedestrian_state.py:40]: Done

File: robot_sf/nav/occupancy.py [hunk 15]
```diff
@@ -218,3 +224,50 @@ def is_in_bounds(self, world_x: float, world_y: float) -> bool:
...
+    def distance_to_robot(self) -> bool:
+        """
+        Gets the euklidean distance to the robot.
+
+        Returns
+        -------
+        float
+            Distance to the robot.
+        """
+        return euclid_dist(self.get_enemy_coords(), self.get_agent_coords())
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/nav/occupancy.py:256]: Done

File: robot_sf/nav/occupancy.py [hunk 16]
```diff
@@ -218,3 +224,50 @@ def is_in_bounds(self, world_x: float, world_y: float) -> bool:
             True if the point is within the bounds of the occupancy, False otherwise.
         """
         return 0 <= world_x <= self.width and 0 <= world_y <= self.height
+
+
+@dataclass
+class EgoPedContinuousOccupancy(ContinuousOccupancy):
+    get_enemy_coords: Callable[[], Vec2D] = None
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/nav/occupancy.py:231]: Done

File: robot_sf/gym_env/env_util.py [hunk 17]
```diff
@@ -90,23 +91,134 @@ def init_spaces(env_config: EnvSettings, map_def: MapDefinition):
...
+    """
+    action_space_robot, obs_space_robot, orig_obs_space_robot = create_spaces(env_config, map_def, create_robot=True)
+    action_space_ped, obs_space_ped, orig_obs_space_ped = create_spaces(env_config, map_def, create_robot=False)
 
+    # As a list [robot, pedestrian]
     # Return the action space, the extended observation space, and the original
     # observation space
-    return action_space, observation_space, orig_obs_space
+    return [action_space_robot, action_space_ped], [obs_space_robot, obs_space_ped], [orig_obs_space_robot, orig_obs_space_ped]
+
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/gym_env/env_util.py:152]: Done

File: robot_sf/sim/simulator.py [hunk 18]
```diff
@@ -166,3 +174,122 @@ def init_simulators(
...
+
+    # Create the pedestrian for this simulator
+    sim_ped = env_config.pedestrian_factory()
+
+        # Create the simulator with the robots and add it to the list
+    sim = PedSimulator(
+        env_config.sim_config, map_def, [sim_robot],
+        goal_proximity, random_start_pos, ego_ped=sim_ped)
+
+    return [sim]
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/sim/simulator.py:295]: Removed num_robots

File: robot_sf/sensor/range_sensor.py [hunk 19]
```diff
@@ -281,6 +281,20 @@ def raycast(scanner_pos: Vec2D, obstacles: np.ndarray, max_scan_range: float,
...
+    """Cast rays in the directions of all given angles outgoing from
+    the scanner's position and detect the minimal collision distance
+    with either a pedestrian or an obstacle (or in case there's no collision,
+    just return the maximum scan range)."""
+    out_ranges = np.full((ray_angles.shape[0]), np.inf)
+    raycast_pedestrians(out_ranges, scanner_pos, max_scan_range, ped_pos, ped_radius, ray_angles)
+    raycast_pedestrians(out_ranges, scanner_pos, max_scan_range, enemy_pos, enemy_radius, ray_angles)
+    raycast_obstacles(out_ranges, scanner_pos, obstacles, ray_angles)
+    return out_ranges
+
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/sensor/range_sensor.py:297]: Done

File: robot_sf/gym_env/pedestrian_env.py [hunk 20]
```diff
@@ -0,0 +1,324 @@
+"""
+`pedestrian_env.py` is a module that defines the simulation environment for a pedestrian.
+It includes classes and protocols for defining the pedetsrians's state, actions, and 
+observations within the environment. 
+
+`PedestrianEnv`: A class that represents the pedestrian's environment. It inherits from `VectorEnv` #TODO: check this
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/gym_env/pedestrian_env.py:6]: Done

File: robot_sf/gym_env/pedestrian_env.py [hunk 21]
```diff
@@ -0,0 +1,324 @@
...
+
+class PedestrianEnv(Env):
+    """
+    Representing a Gymnasium environment for training a adversial pedestrian
+    with reinforcement learning.
+    """
+
+    def __init__(
+            self,
+            env_config: PedEnvSettings = PedEnvSettings(),
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/gym_env/pedestrian_env.py:52]: Done

File: robot_sf/gym_env/pedestrian_env.py [hunk 22]
```diff
@@ -0,0 +1,324 @@
...
+        Returns:
+        - obs: Observation after taking the action.
+        - reward: Calculated reward for the taken action.
+        - term: Boolean indicating if the episode has terminated.
+        - info: Additional information as dictionary.
+        """
+        action_ped = action
+
+        # Process the action through the simulator
+        action_robot, _ = self.robot_model.predict(self.last_obs_robot, deterministic=True)
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/gym_env/pedestrian_env.py:166]: Done

File: robot_sf/render/sim_view.py [hunk 23]
```diff
@@ -269,6 +290,14 @@ def _draw_robot(self, pose: RobotPose):
...
             self._scale_tuple(pose[0]),
             self.robot_radius * self.scaling)
+        
+    def _draw_ego_ped(self, pose: PedPose):
+        # TODO: display robot with an image instead of a circle
+        pygame.draw.circle(
+            self.screen,
+            EGO_PED_COLOR,
+            self._scale_tuple(pose[0]),
+            self.robot_radius * self.scaling)
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/render/sim_view.py:300]: Done

File: robot_sf/gym_env/env_util.py [hunk 24]
```diff
@@ -90,23 +91,133 @@ def init_spaces(env_config: EnvSettings, map_def: MapDefinition):
...
+        agent.observation_space,
         target_sensor_space(map_def.max_target_dist),
         lidar_sensor_space(
             env_config.lidar_config.num_rays,
             env_config.lidar_config.max_scan_dist)
         )
-
-    # Return the action space, the extended observation space, and the original
-    # observation space
     return action_space, observation_space, orig_obs_space
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/gym_env/env_util.py:119]: Done

File: robot_sf/sim/simulator.py [hunk 25]
```diff
@@ -166,3 +174,121 @@ def init_simulators(
...
+
+            new_x = x + distance * cos(angle)
+            new_y = y + distance * sin(angle)
+            if not self.is_obstacle_collision(new_x, new_y):
+                return new_x, new_y
+
+        logger.warning("Could not find a valid proximity point: {fixed_point}.")
+        spawn_id = sample(self.map_def.ped_spawn_zones, k=1)[0] # Spawn in pedestrian spawn_zone
+        initial_spawn = sample_zone(spawn_id, 1)[0]
+        return initial_spawn
```
[the remark this answers is not in the document]
Comment (JuliusMiller) [reply robot_sf/sim/simulator.py:245]: Done
